### ISSUE TITLE
Cannot override list config property in Quarkus 3.17

### ISSUE THREAD

**Issue by fire2** [quarkusio/quarkus#44891]
### Describe the bug

In Quarkus 3.17, a config property of type list specified in yaml as: 

```
countries:
  - de
```

cannot be overwritten via system property such as:

```
mvn compile quarkus:dev -Dcountries=nl
```

This worked fine in 3.16.x and should work according to the [Quarkus documentation](https://quarkus.io/guides/config-reference#system-properties).

### Expected behavior

List config property can be overwritten as on Quarkus 3.16

### Actual behavior

Default value from application.yaml is used instead of value specified in system property. The same behaviour is observed with environment variable, it doesn't override the default value either.

### How to Reproduce?

1. Clone my minimal demo: https://github.com/fire2/quarkus-bug-config-list
2. Follow the steps on the readme to see breaking change (bug) in 3.17 compared to 3.16.

### Output of `uname -a` or `ver`

Darwin LPTP2036320 23.6.0 Darwin Kernel Version 23.6.0: Thu Sep 12 23:36:12 PDT 2024; root:xnu-10063.141.1.701.1~1/RELEASE_ARM64_T6020 arm64

### Output of `java -version`

openjdk version "21.0.1" 2023-10-17 LTS

### Quarkus version or git rev

3.17.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.6

**Comment by radcortez** [comment]
@fire2 can you please check if https://github.com/quarkusio/quarkus/pull/44838 fixes the issue?

Alternatively, you can try: https://github.com/quarkusio/quarkus/issues/44811#issuecomment-2506863705

**Comment by fire2** [comment]
Explicitly using `io.smallrye.config.smallrye-config-source-yaml:3.10.2` with Quarkus 3.17.2 as below didn't solve the issue:

```
<dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>${quarkus.platform.group-id}</groupId>
                <artifactId>${quarkus.platform.artifact-id}</artifactId>
                <version>3.17.2</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
            <dependency>
                <groupId>io.smallrye.config</groupId>
                <artifactId>smallrye-config-source-yaml</artifactId>
                <version>3.10.2</version>
            </dependency>
        </dependencies>
    </dependencyManagement>
```

I confirmed checking the dependency tree that 3.10.2 is being used.

**Comment by radcortez** [comment]
It has to be `smallrye-config-source-yaml:3.9.x`, like `3.9.1`.

**Comment by fire2** [comment]
I thought 3.10.2 has the reverted change. Anyways, I tried now with 3.9.1. 

```
<dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>${quarkus.platform.group-id}</groupId>
                <artifactId>${quarkus.platform.artifact-id}</artifactId>
                <version>3.17.2</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
            <dependency>
                <groupId>io.smallrye.config</groupId>
                <artifactId>smallrye-config-source-yaml</artifactId>
                <version>3.9.1</version>
            </dependency>
        </dependencies>
</dependencyManagement>
```

And it still doesn't work.

**Comment by gsmet** [comment]
@radcortez FWIW, there's a reproducer attached so you should be able to reduce the feedback loop :).

**Comment by radcortez** [comment]
Yes. I was already looking into it. Because the issue seems very similar to #44811, I thought that was the problem. I'm sorry for that.

This is a different case caused by https://github.com/smallrye/smallrye-config/pull/1202/. A collection configuration value can be represented as `list=foo,bar` or `list[0]=foo,list[1]=bar`. While the first version is convenient, it only really works for simple type elements and not for composed types. The API became a bit inconsistent, so we tried to make things a bit more consistent by prioritizing indexed names when populating collections. (BTW, the indexed and the APIs for collections are a SmallRye Config exclusive feature; MicroProfile Config only deals with simple collection elements and the comma-separated values configuration).

This usually is not a problem because users use only one of the styles, but the YAML source is tricky (also related to #44811). The `ConfigSource` API forces us to translate a YAML file content to property-base content.  In reality, the correct representation of a collection in YAML is the indexed form `name[i]`. The `YAMLConfigSource` predates the support of indexed properties, so we indeed flattened and transformed such cases to the comma-separated values (which I wanted to remove in https://github.com/smallrye/smallrye-config/pull/1203.

So, in the end, the `YAMLConfigSource` provides the indexed format and the comma-separated values format. Because the indexed has priority, to be able to override it, you need to use that format instead (`-Dcountries[0]=nl`).

I think I have an idea to support the new priority and still make it overrideable with both styles.

**Comment by fire2** [comment]
is there a target Quarkus version for the fix?

**Comment by radcortez** [comment]
Hopefully in the next release.

**Comment by fire2** [comment]
Hi @radcortez ,
I just tried this on Quarkus 3.18.1 and the issue persists. You can try with my demo (https://github.com/fire2/quarkus-bug-config-list), just change the Quarkus version property. Wasn't it supposed to be fixed on 3.18.x ?

**Comment by fire2** [comment]
@gsmet  can we reopen, please?

**Comment by radcortez** [comment]
Yes, this fix only fixed the programmatic APIs in `SmallRyeConfig`, which I thought was going to be enough to fix the issue. For instance, it works fine if you use `@ConfigMapping`.

The `@ConfigProperty` injection code, is older and a bit different, before we had many of the required APIs in `SmallRyeConfig`, so it also needed the fix: https://github.com/smallrye/smallrye-config/pull/1301

It was merged in Quarkus with: https://github.com/quarkusio/quarkus/pull/46032

**Comment by fire2** [comment]
I see, since it was reported again with https://github.com/quarkusio/quarkus/issues/46016 ... ok, I will try again when Quarkus 3.18.2 is released. Thanks!

**Comment by radcortez** [comment]
Yes, sorry for the inconvenience. Let me know if that works for you.

**Comment by fire2** [comment]
I can confirm the issue is solved with Quarkus 3.18.2. Thanks!
